File: python/day_10.py
from statistics import median

# define functions
def syntax_err(x, part_1 = True):

    illegal = False
    expect = []
    lookup = ['[','{','(','<']
    requires = [']','}',')','>']

    for ii in range(len(x)):
        if x[ii] not in [']','}',')','>']:
            tmp_idx = lookup.index(x[ii])
            expect.append(requires[tmp_idx])
        elif x[ii] == expect[len(expect)-1]:
            del expect[-1]
            continue
        else:
            illegal = True
            break
    
    if part_1:
        if illegal:
            return x[ii]
        else:
            return
    elif not illegal:
        return expect
    else:
        return

def auto_score(x):

    score = 0

    lookup = {
        ")": 1,
        "]": 2,
        "}": 3,
        ">": 4
    }

    x = [lookup[i] for i in x]

    for ii in x:
        score = score * 5
        score += ii

    return score

# import data
f = open("day_10.txt", "r")

input = f.read().splitlines()

f.close()

# process part 1
illegal_chars = []

for ii in input:
    illegal_chars.append(syntax_err(ii))

illegal_chars = list(filter(None, illegal_chars))

scores = {
    ")": 3,
    "]": 57,
    "}": 1197,
    ">": 25137
}

illegal_chars = [scores[x] for x in illegal_chars]

# process part 2
auto_comp = []

for ii in input:
    auto_comp.append(syntax_err(ii, part_1=False))

auto_comp = list(filter(None, auto_comp))

for ii in range(len(auto_comp)):
    auto_comp[ii].reverse()

ac_scores = []

for ii in auto_comp:
    ac_scores.append(auto_score(ii))

# Part 1 answer
print(sum(illegal_chars))

# Part 2 answer
print(median(ac_scores))

File: R/day_10.R
input <- readLines("day_10.txt", warn = FALSE)

.ParseString <- function(input) unlist(strsplit(input, ""))
.SwitchParen <- function (x) return(switch(x, ")" = "(", "]" = "[", ">" = "<", "}" = "{"))

Validate <- function(input) {
  stack <- NULL

  illegal <- NA
  
  for(ii in .ParseString(input)){
    if (
      ii == "(" | ii == "[" | ii == "<" | ii == "{"
    ) {
      stack <- c(stack, ii)
    } else if (stack[length(stack)] == .SwitchParen(ii)) {
      stack <- stack[1:(length(stack) - 1)]
    } else {
      illegal <- ii
      break
    }
  }
  return(illegal)
}

Complete <- function(input) {
  stack <- NULL
  
  for(ii in .ParseString(input)){
    if (
      ii == "(" | ii == "[" | ii == "<" | ii == "{"
    ) {
      stack <- c(stack, ii)
    } else if (stack[length(stack)] == .SwitchParen(ii)) {
      stack <- stack <- if (length(stack) == 1) {NULL} else {stack[1:(length(stack) - 1)]}
    }
  }
  return(stack)
}

Score <- function(input) {
  score <- 0
  for (ii in seq_along(input)) {
    score <- (score * 5) + switch(input[ii], "(" = 1, "[" = 2, "<" = 4, "{" = 3)
  }
  return(score)
}

# Part 1
illegals <- NULL

for (ii in seq_along(input)) {
  check <- Validate(input[ii])
  if (!is.na(check)) illegals <- c(illegals, check)
}

# Part 1 answer
(sum(")" == illegals) * 3) + 
  (sum("]" == illegals) * 57) +
  (sum("}" == illegals) * 1197) +
  (sum(">" == illegals) * 25137)
  
# Part 2
incomplete <- rep(FALSE, length(input))

for (ii in seq_along(input)) {
  if (is.na(Validate(input[ii]))) incomplete[ii] <- TRUE
}

input <- input[incomplete]

calc_scores <- rep(0, length(input))

for (ii in seq_along(input)) {
  calc_scores[ii] <- Score(rev(Complete(.ParseString(input[ii]))))
}

# Part 2 answer
median(calc_scores)
